ext/drpttracer/drpttracer.h

```diff
@@ -0,0 +1,211 @@
+ * sideband config.
+ */
+typedef struct __attribute__((__packed__)) _pt_metadata_t {
+    /** The CPU family. */
+    uint16_t cpu_family;
+    /** The CPU mode. */
+    uint8_t cpu_model;
+    /** The CPU stepping. */
+    uint8_t cpu_stepping;
+    /** perf_event_mmap_page.time_shift */
+    uint16_t time_shift;
+    /** perf_event_mmap_page.time_mult */
+    uint32_t time_mult;
+    /** perf_event_mmap_page.time_zero */
+    uint64_t time_zero;
+} pt_metadata_t;
+
+/***************************************************************************
+ * INIT
+ */
+
+DR_EXPORT
+/**
+ * Initializes the drpttracer extension.  Must be called prior to any of the
+ * other routines.  Can be called multiple times (by separate components,
+ * normally) but each call must be paired with a corresponding call to
+ * drpttracer_exit().
+ *
+ * \return whether successful.
+ */
+bool
+drpttracer_init(void);
+
+DR_EXPORT
+/**
+ * Cleans up the drpttracer extension.
+ */
+void
+drpttracer_exit(void);
+
+/***************************************************************************
+ * PT TRACING APIS
+ */
+
+/**
+ * Success code for each drpttracer operation.
+ */
+typedef enum {
+    /** Operation succeeded. */
+    DRPTTRACER_SUCCESS,
+    /** Operation failed. */
+    DRPTTRACER_ERROR,
+    /** Operation failed: invalid tracing handle. */
+    DRPTTRACER_ERROR_INVALID_TRACING_HANDLE,
+    /** Operation failed: invalid tracing mode. */
+    DRPTTRACER_ERROR_INVALID_TRACING_MODE,
+    /** Operation failed: failed to open perf event. */
+    DRPTTRACER_ERROR_FAILED_TO_OPEN_PERF_EVENT,
+    /** Operation failed: failed to create tracer handle. */
+    DRPTTRACER_ERROR_FAILED_TO_CREATE_PTTRACER_HANDLE,
+    /** Operation failed: failed to start tracing. */
+    DRPTTRACER_ERROR_FAILED_TO_START_TRACING,
+    /** Operation failed: failed to stop tracing. */
+    DRPTTRACER_ERROR_FAILED_TO_STOP_TRACING
+} drpttracer_status_t;
+
+/**
+ * The tracing modes for drpttracer_start_tracing().
+ */
+typedef enum {
+    /* drpttracer only traces in user space. */
+    DRPTTRACER_TRACING_ONLY_USER,
+    /* drpttracer only traces in kernel. */
+    DRPTTRACER_TRACING_ONLY_KERNEL,
+    /* drpttracer traces both user and kernel. */
+    DRPTTRACER_TRACING_USER_AND_KERNEL
+} drpttracer_tracing_mode_t;
+
+DR_EXPORT
+/**
+ * Starts PT tracing. Must be called after drpttracer_init() and before drpttracer_exit().
+ *
+ * \param[in] drcontext  The context of DynamoRIO.
+ * \param[in] mode  The tracing mode.
+ * \param[out] tracer_handle  The tracer handle.
+ *
+ * \note Each tracing corresponds to a trace_handle. When calling
+ * drpttracer_start_tracing(), the caller will get a tracer_handle. And the caller can
+ * stop tracing by passing it to drpttracer_stop_tracing().
+ *
+ * \return the status code.
+ */
+drpttracer_status_t
+drpttracer_start_tracing(IN void *drcontext, IN drpttracer_tracing_mode_t mode,
```

Comment:
How does the user set any parameters like the size of the buffers for PT to use?

Reply:
Done.